java-manta-client/src/main/java/com/joyent/manta/client/MantaClient.java

```diff
@@ -278,17 +286,27 @@ public void deleteRecursive(final String path) throws IOException {
          */
         boolean pathExists = existsAndIsAccessible(path);
 
-        if (pathExists && isDirectoryEmpty(path)) {
+        if (!pathExists) {
+            LOG.debug("Path {} doesn't exist. Finished.", path);
+            return;
+        }
+
+        try {
             this.delete(path);
-        } else if (pathExists) {
-            try {
-                final int waitTime = 400;
-                Thread.sleep(waitTime);
-                LOG.warn("First attempt to delete directory failed, retrying");
-                // Re-attempt to delete the directory
-                this.deleteRecursive(path);
-            } catch (InterruptedException ie) {
-                // We don't need to do anything, just exit
+        } catch (MantaClientHttpResponseException e) {
+            // Directory wasn't empty
+            if (e.getServerCode().equals(DIRECTORY_NOT_EMPTY_ERROR)) {
+                try {
+                    final int waitTime = 400;
```

Comment:
Sometimes we hit a race condition when recursively deleting. Many times, there is another recursive delete operation also running. Giving it a moment to wait for the other process to stop increases the likelihood that the retry succeeds. When you reason about it theoretically, this makes no sense. When you look at how all of the SDKs are implemented and how most users implement them, they tend to fall in this pattern and this becomes a pragmatic choice.
Honestly, this design decision needs review and revisiting, so I've created [this issue](https://github.com/joyent/java-manta/issues/125).